Add support for both Material UI 4 and 5 in @rjsf/material-ui (fix #2555)

heath-freenome opened the pull request:
### Reasons for making this change

Fix #2555 by adding side-by-side support for both the Material UI 4 and 5 themes as follows:
- Updated `core` to add React 17 as a peer-dependency to support the `@mui` libraries which requires it
  - This required removing the `create-react-ref` library that didn't support React 17, updating the single usage of it in a test with `React.createRef()`
- Ran the [mui v4->v5](https://mui.com/guides/migration-v4/) over the `@material-ui` source code
  - Cleaned up the imports structures of the code base
  - Removed the `TextFieldProps` in favor of simply using the updated `WidgetProps` in may of the `TextField` based widgets
  - Made a few other smaller Typescript fixes to a few Widgets
  - Created a new React context, `MuiComponentContext` that will be used for encapsulating the`@material-ui` or `@mui` components and icons
    - Updated ALL of the field, widget and helper components in `@rjsf/material-ui` to use the `MuiComponentContext` to obtain the appropriate Material UI definitions
      - Removed the  `@rjsf/material-ui` `IconButton` component in favor of explicitly using the Material UI `IconButton` wrapped around each of the 4 Icons (`Add`, `ArrowDownward`, `ArrowUpward` or `Remove`)
  - Updated the existing Material UI 4 `Theme` to attempt to load a `MaterialUIContext` from the `@material-ui` libraries, which is used in the new `Mui4TagName` component to wrap a `form` with a `MuiComponentContext` for Material UI 4
  - Updated the material-ui version 4 tests snapshot
  - Added new `MuiForm5` and `Theme5` components that attempt to load a `Mui5Context` from the `@mui` libraries, which is used in the new `Mui5TagName` component to wrap a `form` with a `MuiComponentContext` for Material UI 5
  - Added `mui5` tests which are a copy of the version 4 tests except that one failing test was skipped (due to [mui issue](https://github.com/mui-org/material-ui/issues/29632)
  - Updated the root and `material-ui` `README.md` and a few other theme related documentation files
  - Added a `screenshot5.png` to the root directory
- Updated `playground` to bump the `@material-ui` versions and as well as the `babel` ecosystem and `core-js`
  - Added a new `src/emptyLibrary.js` file that simply returns null
  - Updated the webpack config files to add new resolve aliases for the `@mui` libraries to return `emptyLibrary` since trying to load them will break things without `React 17`
- Added a new `playground-r17` package that contains only the `core` and `material-ui` themes
  - This required upgrading `react-monoco-editor`, `monoco-editor` and `monaco-editor-webpack-plugin` to versions that support React 17
  - Updated `DemoFrame` to use the `@emotion/react` `CacheProvider` in addition to the older `@material-ui/styles` `StylesProvider`
  - NOTE: I am not sure how to update the main deployment system to integrate this with the main published playground
    - @epicface perhaps we can work together to figure this out?
    - In the meantime, developers can use this non-published playground to test with
 
### Checklist

* [x] **I'm updating documentation**
  - [x] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
  - [x] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/master/CHANGELOG.md) with a description of the PR
* [x] **I'm adding a new feature**
  - [x] I've updated the playground with an example use of the feature

epicfaace commented:
Thanks for the PR! So, given this PR, if a user wants to use mui 4, how would they do that?

epicfaace commented:
Oh, I see the release notes, I get it now.

heath-freenome commented:
> Thanks for the PR! So, given this PR, if a user wants to use mui 4, how would they do that?

Answering for others... For backwards compatibility purposes, the default export is still the `MuiForm` and the `MuiForm` and `Theme` objects are still for Material UI 4. There are also the `MuiForm4` and `Theme4` aliases to those components. To use Material UI 5, one would have to choose to use the `MuiForm5` or `Theme5` components.

heath-freenome commented:
@epicfaace Did you want me to do a walk-thru with you on this? I can set up a Zoom meeting

epicfaace commented:
Yes, that would be great!
https://calendly.com/aramaswamis/meeting?month=2021-11

--
Ashwin Ramaswami


On Tue, Nov 16, 2021 at 7:37 PM Heath C ***@***.***> wrote:

> @epicfaace <https://github.com/epicfaace> Did you want me to do a
> walk-thru with you on this? I can set up a Zoom meeting
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rjsf-team/react-jsonschema-form/pull/2605#issuecomment-970939438>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM4MXYUBSXAMLMB35GDI6DUML2NJANCNFSM5IFCBBYQ>
> .
>

sforehand commented:
hi. first, thanks for all your awesome work!  Question for you... we are currently working thru the upgrade of our code from v4 to v5 MUI.  I am curious what you believe your timeline is to release v5 support?  Certainly not trying to hurry you but just curious so I can plan accordingly.  Thanks so much!

heath-freenome commented:
@epicfaace If I resolve the conflicts is the PR ready to merge?

kerematam commented:
Hi, @heath-freenome great work!

Do we have any updates on this PR?

CC: @epicfaace

heath-freenome commented:
I’m going to rebase to master on Monday, resolve conflicts and hope that
@epicfaace completes the review

On Wed, Nov 24, 2021 at 8:00 AM katam ***@***.***> wrote:

> Hi, @heath-freenome <https://github.com/heath-freenome> great work!
>
> Do we have any updates on this PR?
>
> CC: @epicfaace <https://github.com/epicfaace>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/rjsf-team/react-jsonschema-form/pull/2605#issuecomment-977958522>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMKJCZFZM5TE44YIADDL6FLUNT4XNANCNFSM5IFCBBYQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

-- 
This e-mail is private and confidential and is for the addressee only. If 
misdirected, please notify us by telephone, confirming that it has been 
deleted from your system and any hard copies destroyed. You are strictly 
prohibited from using, printing, distributing or disseminating it or any 
information contained in it save to the intended recipient.

epicfaace commented:
I'll take a look, thanks

--
Ashwin Ramaswami


On Sat, Nov 27, 2021 at 11:55 AM Heath C ***@***.***> wrote:

> I’m going to rebase to master on Monday, resolve conflicts and hope that
> @epicfaace completes the review
>
> On Wed, Nov 24, 2021 at 8:00 AM katam ***@***.***> wrote:
>
> > Hi, @heath-freenome <https://github.com/heath-freenome> great work!
> >
> > Do we have any updates on this PR?
> >
> > CC: @epicfaace <https://github.com/epicfaace>
> >
> > —
> > You are receiving this because you were mentioned.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/rjsf-team/react-jsonschema-form/pull/2605#issuecomment-977958522
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AMKJCZFZM5TE44YIADDL6FLUNT4XNANCNFSM5IFCBBYQ
> >
> > .
> > Triage notifications on the go with GitHub Mobile for iOS
> > <
> https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675
> >
> > or Android
> > <
> https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub
> >.
> >
> >
>
> --
> This e-mail is private and confidential and is for the addressee only. If
> misdirected, please notify us by telephone, confirming that it has been
> deleted from your system and any hard copies destroyed. You are strictly
> prohibited from using, printing, distributing or disseminating it or any
> information contained in it save to the intended recipient.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rjsf-team/react-jsonschema-form/pull/2605#issuecomment-980704784>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM4MXZ3PN5BELEL3GVLDDLUOEERFANCNFSM5IFCBBYQ>
> .
>

heath-freenome commented:
@epicfaace I've rebased on top of master and dealt with the merge conflicts

heath-freenome commented:
@epicfaace If it would be easier for us to get on another call I'm happy to do it. I'd love to get this feature merged

epicfaace commented:
I'm a bit backed up until Dec 20 but will take a look then

--
Ashwin Ramaswami


On Fri, Dec 3, 2021 at 1:11 PM Heath C ***@***.***> wrote:

> @epicfaace <https://github.com/epicfaace> If it would be easier for us to
> get on another call I'm happy to do it. I'd love to get this feature merged
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rjsf-team/react-jsonschema-form/pull/2605#issuecomment-985726632>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM4MX53FSOZAUQFVQGXEETUPEB5PANCNFSM5IFCBBYQ>
> .
>

JeremyRubin commented:
as a small note / priority escallation, within a single project it seems to be not possible to have material-ui (v4) and v5 both being used, which is quite unfortunate (the styling apparantly conflicts and breaks components like Menu).

JeremyRubin commented:
yep, totally understandable and I agree with that.

it can sometimes be untennable if one of your dependencies (e.g., perhaps a library that adds a settings wizard) relies on one and your project relies on the other.

I think this is probably more of an issue for the material project to fix than for you.

For now, in my project, i've just gone to the unstyled default until this patch lands.

heath-freenome commented:
> I'm a bit backed up until Dec 20 but will take a look then

Hoping to see some feedback soon!

epicfaace commented:
Yes

--
Ashwin Ramaswami


On Mon, Dec 20, 2021 at 6:50 PM Heath C ***@***.***> wrote:

> I'm a bit backed up until Dec 20 but will take a look then
>
> Hoping to see some feedback soon!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rjsf-team/react-jsonschema-form/pull/2605#issuecomment-998353925>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM4MXYOPDLDPZWFA6MPEXLUR66MTANCNFSM5IFCBBYQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

heath-freenome commented:
Awesome! If you want to go over it again together I'm free for much of tomorrow. Then I go on vacation until 1/4... It would be great if we could get it merged in before the new year... but if not we can keep em waiting :)

heath-freenome commented:
@epicfaace Hope you had a nice end of 2021, beginning of 2022... Let me know if you have any questions and/or want to walk through this again

epicfaace commented:
Yes

--
Ashwin Ramaswami


On Tue, Jan 4, 2022 at 1:04 PM Heath C ***@***.***> wrote:

> @epicfaace <https://github.com/epicfaace> Hope you had a nice end of
> 2021, beginning of 2022... Let me know if you have any questions and/or
> want to walk through this again
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rjsf-team/react-jsonschema-form/pull/2605#issuecomment-1005049544>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM4MX7JVHYSSEHW36F4YEDUUMZCZANCNFSM5IFCBBYQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

heath-freenome commented:
> Yes
> […](#)
> -- Ashwin Ramaswami
> On Tue, Jan 4, 2022 at 1:04 PM Heath C ***@***.***> wrote: @epicfaace <https://github.com/epicfaace> Hope you had a nice end of 2021, beginning of 2022... Let me know if you have any questions and/or want to walk through this again — Reply to this email directly, view it on GitHub <[#2605 (comment)](https://github.com/rjsf-team/react-jsonschema-form/pull/2605#issuecomment-1005049544)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAM4MX7JVHYSSEHW36F4YEDUUMZCZANCNFSM5IFCBBYQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

I'm going you assume you'd like a walk-thru again... Shall I email you a zoom meeting?

epicfaace commented:
Sorry, that's not what I meant.

I think the best way to speed up my review is -- if you can, do you mind deploying both playgrounds somewhere so I can see them without needing to run rjsf locally? Thanks.

heath-freenome commented:
> Sorry, that's not what I meant.
> 
> I think the best way to speed up my review is -- if you can, do you mind deploying both playgrounds somewhere so I can see them without needing to run rjsf locally? Thanks.

I'm not sure how to set up github pages on my fork. Also, you and I need to figure out how to deploy both playgrounds for the production release of @rjsf

epicfaace commented:
> Also, you and I need to figure out how to deploy both playgrounds for the production release of @rjsf

I think we mentioned in our last call -- potentially that when someone chooses "material ui 5" from the dropdown, then it would switch to the other playground?

heath-freenome commented:
> > Also, you and I need to figure out how to deploy both playgrounds for the production release of @rjsf
> 
> I think we mentioned in our last call -- potentially that when someone chooses "material ui 5" from the dropdown, then it would switch to the other playground?

I looked into trying to do that and it seemed very complicated... I think just having two static links for now would be better... Have them switch back and forth. And I feel like doing that in a follow-on PR once this is merged is probably a better choice

epicfaace commented:
Ok, sounds good. I just need to find some time to run this locally and
inspect it. I will do it when I get time, please keep pinging me if I don't
get to it

--
Ashwin Ramaswami


On Thu, Jan 6, 2022 at 7:28 PM Heath C ***@***.***> wrote:

> Also, you and I need to figure out how to deploy both playgrounds for the
> production release of @rjsf <https://github.com/rjsf>
>
> I think we mentioned in our last call -- potentially that when someone
> chooses "material ui 5" from the dropdown, then it would switch to the
> other playground?
>
> I looked into trying to do that and it seemed very complicated... I think
> just having two static links for now would be better... Have them switch
> back and forth. And I feel like doing that in a follow-on PR once this is
> merged is probably a better choice
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rjsf-team/react-jsonschema-form/pull/2605#issuecomment-1007041040>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM4MX2RL5NBD2J467JXNCDUUYXSHANCNFSM5IFCBBYQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

heath-freenome commented:
@epicfaace Ok, I've rebased this PR onto the latest master. In the process I cleaned up a couple of things and had to redo the `material-ui` snapshots. Hopefully you can get a review on this soon!  Thanks for your time, I know you are busy

epicfaace commented:
@heath-freenome I made a change -- see https://github.com/heath-freenome/react-jsonschema-form/pull/1

epicfaace commented on packages/core/test/Form_test.js:
```diff
@@ -1,10 +1,9 @@
 import { renderIntoDocument, Simulate } from "react-dom/test-utils";
 import { render, findDOMNode } from "react-dom";
 import { Portal } from "react-portal";
-import { createRef } from "create-react-ref";
```
is this a breaking change? are we changing the minimum version of react we support now?

epicfaace replied:
if so, we should mention this in the changelog

epicfaace replied:
and update package.json accordingly

heath-freenome replied:
I don't believe this is a breaking change... This is a test and we are on a version of react that has `createRef` already provided.

heath-freenome replied:
Although you may be correct about the minimum version of react in the core package.json likely needing to be bumped

epicfaace replied:
So this means that it's a breaking change as the minimum version of react required to use rjsf is now react 16.3 (as opposed to 16.0)

epicfaace reviewed (commented):
lgtm, I think we'll go ahead and make this a major version bump, are you ok with that?

heath-freenome commented:
> @heath-freenome I made a change -- see [heath-freenome#1](https://github.com/heath-freenome/react-jsonschema-form/pull/1)

Thanks for fixing the dual playground issue! Now the main playground has material-ui 5

heath-freenome commented:
> lgtm, I think we'll go ahead and make this a major version bump, are you ok with that?

As the maintainer, you own that call... The code itself only has a potential breaking change in that we bumped the material-ui version requirements... I'm fine with bumping the major version

epicfaace commented:
> > lgtm, I think we'll go ahead and make this a major version bump, are you ok with that?
> 
> As the maintainer, you own that call... The code itself only has a potential breaking change in that we bumped the material-ui version requirements... I'm fine with bumping the major version although I don't believe it needs it

In your opinion, how significant are the snapshot changes for the material-ui-4 theme?

heath-freenome commented:
> > > lgtm, I think we'll go ahead and make this a major version bump, are you ok with that?
> > 
> > 
> > As the maintainer, you own that call... The code itself only has a potential breaking change in that we bumped the material-ui version requirements... I'm fine with bumping the major version although I don't believe it needs it
> 
> In your opinion, how significant are the snapshot changes for the material-ui-4 theme?

Primarily the changes are in the classNames... There also seem to be a few minor changes to the props passed to SVGs and some of the controls. Overall, the structure is the same. I'd rate the significance on a scale of 0 (no changes) to 10 (totally changed) to be about a 2.5. Are you still considering the major vs minor bump? There are a few changes in the PRs list related to AVJ changes that ARE breaking changes (I'd love to get that error message localization change in)... If we bundle them with this change then I would definitely agree with a major bump

epicfaace commented:
This should also fix https://github.com/rjsf-team/react-jsonschema-form/issues/2517

heath-freenome commented:
> This should also fix #2517

Yup!

heath-freenome commented:
@epicfaace Can we get this merged soon? Is it just a decision about minor vs major bump?

epicfaace commented:
At the moment, I'm thinking -- let's bump the major version because:
- min supported React version changed from 16 to 16.3
- min supported material-ui 4 version changed (is that right? do you know what the min version is now?)

We need to update the changelog accordingly and then let's merge.

epicfaace commented:
@heath-freenome I'm also thinking we should set up a weekly time / "rjsf work session" where people join a call and can talk about PRs, issues, releases, etc. -- might make it easier to set a cadence for rjsf work and make progress on PRs such as this. wdyt?

heath-freenome commented:
> At the moment, I'm thinking -- let's bump the major version because:
> 
> * min supported React version changed from 16 to 16.3
> * min supported material-ui 4 version changed (is that right? do you know what the min version is now?)
> 
> We need to update the changelog accordingly and then let's merge.

@epicfaace changelog updated! Let's get this out to people... Also, maybe add a few more breaking changes before pushing the new release as well like:
  - https://github.com/rjsf-team/react-jsonschema-form/pull/2634
  - https://github.com/rjsf-team/react-jsonschema-form/pull/2542

heath-freenome commented:
@epicfaace Hope you had a good weekend... how can I help get this over the finish line

epicfaace commented:
Yes, taking a look soon. thanks!

--
Ashwin Ramaswami


On Mon, Jan 31, 2022 at 9:05 PM Heath C ***@***.***> wrote:

> @epicfaace <https://github.com/epicfaace> Hope you had a good weekend...
> how can I help get this over the finish line
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rjsf-team/react-jsonschema-form/pull/2605#issuecomment-1026408762>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAM4MX4POMR7V6VWCV4V3JLUY45VDANCNFSM5IFCBBYQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

heath-freenome commented:
Rebased to master again

epicfaace commented:
Once we fix the merge, we should be ready to get this in.

heath-freenome commented:
@epicfaace let me know if you have problems with the playground... I believe I got it all working

epicfaace commented:
@heath-freenome I was able to run it, but when I go to the "Numbers" tab on the playground with material-ui-5, the screen becomes blank with this error:

![image](https://user-images.githubusercontent.com/1689183/152588273-84ce404e-fd23-41b3-8107-09b9c1d9d0a9.png)

heath-freenome commented:
> @heath-freenome I was able to run it, but when I go to the "Numbers" tab on the playground with material-ui-5, the screen becomes blank with this error:
> 
> ![image](https://user-images.githubusercontent.com/1689183/152588273-84ce404e-fd23-41b3-8107-09b9c1d9d0a9.png)

AHHHH... good catch. I forgot to export `Slider` into the context... fix incoming

heath-freenome commented:
Whoops... I just meant to comment... not close :(

epicfaace commented:
alternate datetime widgets seem broken

![image](https://user-images.githubusercontent.com/1689183/152600106-bdcd2436-054d-4ecc-b50f-de9d8b7364e6.png)

epicfaace commented:
^ actually, let's not fix the above because it's also broken in material-ui-4

epicfaace reviewed (approved):
lgtm

formatlos commented:
is there a release planned for this already?

rollercodester commented:
> is there a release planned for this already?

See https://github.com/rjsf-team/react-jsonschema-form/issues/2694#issuecomment-1034164718
